File: haskell/Problem25.hs
-- Copyright 2015 Mitchell Kember. Subject to the MIT License.
-- Project Euler: Problem 25
-- 1000-digit Fibonacci number

module Problem25 where

solve :: Int
solve = go 1 1 1 1 10
  where
    go _ _ index 1000 _ = index
    go a b index count next
        | a' >= next = go a' b' index' (count + 1) (next * 10)
        | otherwise = go a' b' index' count next
      where
        a' = b
        b' = a + b
        index' = index + 1

File: c++/problem_25.cpp
// Copyright 2016 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 25
// 1000-digit Fibonacci number

#include <gmpxx.h>

namespace problem_25 {

long solve() {
	mpz_class first(1);
	mpz_class second(1);
	mpz_class power_of_ten(10);
	mpz_class* a = &first;
	mpz_class* b = &second;

	long index = 1;
	long digits = 1;
	while (digits < 1000) {
		std::swap(a, b);
		*b += *a;
		++index;
		if (*a / power_of_ten >= 1) {
			++digits;
			power_of_ten *= 10;
		}
	}
	return index;
}

} // namespace problem_25
